pkg/rule/postgres.go

```diff
@@ -0,0 +1,498 @@
+		configID:    raw.ConfigID,
+		createdAt:   raw.CreatedAt,
+		criteria:    &criteria,
+		description: raw.Description,
+		endTime:     &raw.EndTime,
+		kind:        raw.Kind,
+		name:        raw.Name,
+		startTime:   &raw.StartTime,
+		updatedAt:   raw.UpdatedAt,
+	}, nil
+}
+
+func (r *pgRepo) UpdateWith(input rule) (rule, error) {
+	rawBuckets, err := json.Marshal(input.buckets)
+	if err != nil {
+		return rule{}, errors.Wrap(err, "marshal buckets")
+	}
+
+	rawCriteria, err := json.Marshal(input.criteria)
+	if err != nil {
+		return rule{}, errors.Wrap(err, "marshal criteria")
+	}
+
+	_, err = r.db.NamedExec(pgRuleUpdate, map[string]interface{}{
+		"id":          input.id,
+		"active":      input.active,
+		"activatedAt": input.activatedAt,
+		"configId":    input.configID,
+		"buckets":     rawBuckets,
+		"createdAt":   input.createdAt,
+		"criteria":    rawCriteria,
+		"description": input.description,
+		"endTime":     input.endTime,
+		"kind":        input.kind,
+		"name":        input.name,
+		"startTime":   input.startTime,
+		"updatedAt":   input.updatedAt,
+	})
+	if err != nil {
+		switch errors.Cause(pg.Wrap(err)) {
+		case pg.ErrRelationNotFound:
+			if err := r.setup(); err != nil {
+				return rule{}, err
+			}
+
+			return r.UpdateWith(input)
+		case sql.ErrNoRows:
+			return rule{}, errors.Wrap(errors.ErrNotFound, "update rule")
+
+		default:
+			return rule{}, fmt.Errorf("update named exec: %s", err)
+		}
+	}
+
+	return rule{
+		id:          input.id,
+		active:      input.active,
+		activatedAt: input.activatedAt,
+		buckets:     input.buckets,
+		configID:    input.configID,
+		createdAt:   input.createdAt,
+		criteria:    input.criteria,
+		description: input.description,
+		endTime:     input.endTime,
+		kind:        input.kind,
+		name:        input.name,
+		startTime:   input.startTime,
+		updatedAt:   input.updatedAt,
+	}, nil
+}
+
+func (r *pgRepo) ListAll(configID string) ([]rule, error) {
+	rules := []rule{}
+
+	raw := struct {
+		ID          string    `db:"id"`
+		Active      bool      `db:"active"`
+		ActivatedAt time.Time `db:"activated_at"`
+		Buckets     []byte    `db:"buckets"`
+		ConfigID    string    `db:"config_id"`
+		CreatedAt   time.Time `db:"created_at"`
+		Criteria    []byte    `db:"criteria"`
+		Description string    `db:"description"`
+		EndTime     time.Time `db:"end_time"`
+		Kind        kind      `db:"kind"`
+		Name        string    `db:"name"`
+		StartTime   time.Time `db:"start_time"`
+		UpdatedAt   time.Time `db:"updated_at"`
+	}{}
+
+	rows, err := r.db.Queryx(pgRuleListAll)
+	if err != nil {
+		switch errors.Cause(pg.Wrap(err)) {
+		case pg.ErrRelationNotFound:
+			if err := r.setup(); err != nil {
+				return []rule{}, err
+			}
+
+			return r.ListAll(configID)
+		case sql.ErrNoRows:
+			return []rule{}, errors.Wrap(errors.ErrNotFound, "list all rules")
+
+		default:
+			return []rule{}, fmt.Errorf("list all rules: %s", err)
+		}
+	}
+
+	for rows.Next() {
+		err = rows.StructScan(&raw)
```

Comment:
It best be avoided to reuse the same instance for the scan as it might have unforeseen side-effects. It could be easily addressed by moving the construction of `raw` inside the loop.

Reply:
good catch